session.go

```diff
@@ -330,28 +346,37 @@ func (s *Session) keepalive() {
 // waitForSendErr waits to send a header, checking for a potential shutdown
 func (s *Session) waitForSend(hdr header, body io.Reader) error {
 	errCh := make(chan error, 1)
+	defer func() {
+		select {
+		case <-errCh:
+		default:
+		}
+	}()
 	return s.waitForSendErr(hdr, body, errCh)
 }
 
 // waitForSendErr waits to send a header with optional data, checking for a
 // potential shutdown. Since there's the expectation that sends can happen
 // in a timely manner, we enforce the connection write timeout here.
 func (s *Session) waitForSendErr(hdr header, body io.Reader, errCh chan error) error {
-	t := timerPool.Get()
-	timer := t.(*time.Timer)
-	timer.Reset(s.config.ConnectionWriteTimeout)
-	defer func() {
-		timer.Stop()
-		select {
-		case <-timer.C:
-		default:
-		}
-		timerPool.Put(t)
-	}()
-
-	ready := sendReady{Hdr: hdr, Body: body, Err: errCh}
+	// deadline captures the absolute deadline for this write
+	// at the instant it was performed. We have two outcomes to cover:
+	// 1. deadline expiring before we manage to publish to sendCh (buffered),
+	//    controlled by the above timer.
+	// 2. deadline expiring after the above, and during IO. This is controlled
+	//    by the send* methods, which need to know the original deadline.
+	deadline := time.Now().Add(s.config.ConnectionWriteTimeout)
+
+	timer, cancelFn := pooledTimer(s.config.ConnectionWriteTimeout)
+	defer cancelFn()
+
+	ready := sendReady{Hdr: hdr, Body: body, Err: errCh, Deadline: deadline}
 	select {
 	case s.sendCh <- ready:
+		if s.netconn != nil {
+			// if we have a net conn, the write timer has been handed off, so cancel it now.
+			cancelFn()
```

Comment:
> Also, in the netconn case, we still have the original race. We need to make sure we don't return from this function until the send loops tell us they're no longer looking at Body.

`sendConn` only processes one write at a time. If the write is in `sendConn`, the `waitForSendErr` timer has already been cancelled, and we're the ones in charge of returning control. Given that we now set a write deadline, there is no chance that a background write will linger in the background, after we return control. So I think the data race is salvaged. I ran the tests pre-PR and post-PR with `-race` and it confirmed the race was gone.

Reply:
> We can't continue using the timer after we cancel it and put it back in the pool.

Could you point to where we're using it again? It's wrapped in a `sync.Once` (couldn't think of a way to avoid the once).

Reply:
> Could you point to where we're using it again? It's wrapped in a sync.Once (couldn't think of a way to avoid the once).

We're waiting on it in the next select. That is, `cancelFunc` is returning it to the pool but then we wait on it. Someone else might come along and try to use it.

> Given that we now set a write deadline, there is no chance that a background write will linger in the background, after we return control.

The scheduler can still do this. That is:

1. The `Write` in `send` is busy doing some *synchronous* operation (there's no guarantee that it'll obey the deadline to the instant).
2. `waitForSendErr` notices the timeout.
3. `waitForSendErr` returns.
4. The user writes to the buffer.
5. `send`'s `Write` call reads from the buffer.

Reply:
> We're waiting on it in the next select. That is, cancelFunc is returning it to the pool but then we wait on it. Someone else might come along and try to use it.

Damn, this is a merge error. The `case <-timer.C` on the second select should not be there. At that point the timer has been cancelled, and the responsibility has been handed off to `sendConn`. 

> The scheduler can still do this.

Given the `waitForSendErr` timer would've been cancelled, how would it notice the timeout?

Reply:
> Damn, this is a merge error. The case <-timer.C on the second select should not be there. At that point the timer has been cancelled, and the responsibility has been handed off to sendConn.

That only applies for `netconn`, right?

> Given the waitForSendErr timer would've been cancelled, how would it notice the timeout?

Ah. Yes, you're right (for the netconn case, I think).

---

Note: to be fully correct, I believe the second switch needs to use `s.recvDoneCh` instead of `s.shutdownCh`.